src/sentry/spans/consumers/process/factory.py

```diff
@@ -301,38 +186,196 @@ def create_with_partitions(
-            topic=self.output_topic,
-            next_step=NoOp(),
-        )
+        buffer = RedisSpansBufferV2(assigned_shards=[p.index for p in partitions])
 
-        unfold_step = Unfold(generator=expand_segments, next_step=produce_step)
-
-        commit_step = CommitSpanOffsets(commit=commit, next_step=unfold_step)
-
-        batch_processor = RunTask(
-            function=batch_write_to_redis,
-            next_step=commit_step,
+        flusher = SpanFlusher(
+            buffer,
+            self.producer,
+            self.output_topic,
+            self.max_flush_segments,
+            self.max_inflight_segments,
+            next_step=committer,
         )
 
-        batch_step = BatchStep(
+        run_task = run_task_with_multiprocessing(
+            function=partial(process_batch, buffer),
+            next_step=flusher,
             max_batch_size=self.max_batch_size,
             max_batch_time=self.max_batch_time,
-            next_step=batch_processor,
+            pool=self.__pool,
+            input_block_size=self.input_block_size,
+            output_block_size=self.output_block_size,
         )
 
-        return run_task_with_multiprocessing(
-            function=process_message,
-            next_step=batch_step,
+        batch = BatchStep(
             max_batch_size=self.max_batch_size,
             max_batch_time=self.max_batch_time,
-            pool=self.__pool,
-            input_block_size=self.input_block_size,
-            output_block_size=self.output_block_size,
+            next_step=run_task,
         )
 
+        def add_timestamp_cb(message: Message[KafkaPayload]) -> tuple[int, KafkaPayload]:
+            return (
+                int(message.timestamp.timestamp() if message.timestamp else time.time()),
+                message.payload,
+            )
+
+        add_timestamp = RunTask(
+            function=add_timestamp_cb,
+            next_step=batch,
+        )
+
+        return add_timestamp
+
     def shutdown(self) -> None:
         self.producer.close()
         self.__pool.close()
+
+
+def process_batch(
+    buffer: RedisSpansBufferV2, values: Message[ValuesBatch[tuple[int, KafkaPayload]]]
+) -> int:
+    min_timestamp = None
+    spans = []
+    for value in values.payload:
+        timestamp, payload = value.payload
+        if min_timestamp is None or timestamp < min_timestamp:
+            min_timestamp = timestamp
+
+        val = rapidjson.loads(payload.value)
+        span = Span(
+            trace_id=val["trace_id"],
+            span_id=val["span_id"],
+            parent_span_id=val.get("parent_span_id"),
+            project_id=val["project_id"],
+            payload=payload.value,
+            # TODO: validate, this logic may not be complete.
+            is_segment_span=(
+                val.get("parent_span_id") is None
+                or get_path(val, "sentry_tags", "op") == "http.server"
+                or val.get("is_remote")
+            ),
```

Comment:
I would suggest to separate this:
1. Parse JSON into a `Span` instance. We should actually use the codec for this, but you can also keep this manual implementation.
2. Infer `is_segment` using attributes of the instance. This is business logic - I'd suggest to create a dedicated file for this, potentially even outside of the `sentry.spans.consumers` module. I think the buffer / consumer should _use_ this logic. 

That `is_segment` function should get the following parameters injected: `is_remote: bool`, `kind: SpanKind`, `parent_span_id: str`.